quickwit/quickwit-common/src/tower/pool.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::any::TypeId;
+use std::borrow::Borrow;
+use std::cmp::{Eq, PartialEq};
+use std::collections::HashMap;
+use std::fmt;
+use std::hash::Hash;
+use std::sync::Arc;
+
+use futures::{Stream, StreamExt};
+use tokio::sync::RwLock;
+
+use super::Change;
+
+/// A pool of `S` services identified by `K` keys. The pool can be updated manually by calling the
+/// `add/remove` methods or by listening to a stream of changes.
+#[derive(Default)]
+pub struct Pool<K, S> {
+    inner: Arc<RwLock<InnerPool<K, S>>>,
+}
+
+impl<K, S> fmt::Debug for Pool<K, S>
+where
+    K: 'static,
+    S: 'static,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Pool<{:?}, {:?}>", TypeId::of::<K>(), TypeId::of::<S>())
+    }
+}
+
+impl<K, S> Clone for Pool<K, S> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+#[derive(Default)]
+struct InnerPool<K, S> {
+    nodes: HashMap<K, S>,
```

Comment:
This is a pool of services... so why not replacing `nodes` by `services`? 

Same for `get_node` which is a bit strange when you get a service in return. Why not `get_service` or `get`?

If I understand correctly the usage, we will have a pool of services identified by a `node_id` key or a `grpc_address` key.

Reply:
We can stay generic in the naming then, because it can be a pool of anything really:
For the search client, it's probably going to be `Pool<NodeUid, SearchServiceClient>`. For the control plane, it'll be probably a `Pool<NodeUid, (Vec<IndexingTasks>, IndexingServiceClient)>`, for the event broker, it will be `Pool<NodeUid, EventHandlers>`, etc.

Reply:
Yes, I was going to suggest being agnostic about the value types. I like to generic way here.